Recreate stateful set on change in immutable field `volumeClaimTemplates`

matej-g opened the pull request:
Fixes #1491.

Since `volumeClaimTemplates` is an immutable field, this PR provides a solution based on recreation of the stateful set when a change in this field is detected. This is analogous to how other immutable field (selector) is handled.

In addition, this PR simplifies the `VolumeClaimTemplates` method and removes unnecessary parameter.

A unit test has been added. This has also been tested manually.

jaronoff97 commented on .chloggen/1492-fix-stateful-set-reconcile.yaml:
```diff
@@ -0,0 +1,16 @@
+component: "target allocator"
+
+# A brief description of the change. Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Save the scrape config response in the HTTP server upon relevant config change, instead of building it on every handler call. At the same time, this avoids data race when accessing the scrape configs map."
```
I think this changelog entry is incorrect?

matej-g replied:
Looks like I forgot to hit the save button 🤦

matej-g commented on .chloggen/1492-fix-stateful-set-reconcile.yaml:
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
```
Fixed 👍

matej-g commented:
I added more changes, seems like we could not simply compare `volumeClaimTemplates` since some fields are populated automatically by the API and this was causing endless deletion / creation loop. I added a method to compare user-relevant fields by hand. I also added extra E2E test to cover this particular case.

What I'm also wondering if missing stateful set should be considered an error in the [`Self`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/pkg/collector/reconcile/opentelemetry.go#L93) reconciliation task - if we delete the stateful set with intention to recreate it during the next cycle, it missing during present cycle should not be an error.

iblancasa commented on pkg/collector/reconcile/statefulset.go:
```diff
@@ -72,9 +72,10 @@ func expectedStatefulSets(ctx context.Context, params Params, expected []appsv1.
-			params.Log.V(2).Info("Spec.Selector change detected, trying to delete, the new collector statfulset will be created in the next reconcile cycle", "statefulset.name", existing.Name, "statefulset.namespace", existing.Namespace)
+		// Check for immutable fields. If set, we cannot modify the stateful set, otherwise we will face reconciliation error.
+		if needsDeletion, fieldName := hasImmutableFieldChange(&desired, existing); needsDeletion {
+			params.Log.V(2).Info("Immutable field change detected, trying to delete, the new collector statfulset will be created in the next reconcile cycle",
```
```suggestion
			params.Log.V(2).Info("Immutable field change detected, trying to delete, the new collector statefulset will be created in the next reconcile cycle",
```

iblancasa reviewed (changes requested):
Just a minor comment

jaronoff97 commented on pkg/collector/reconcile/statefulset.go:
```diff
@@ -145,3 +146,43 @@ func deleteStatefulSets(ctx context.Context, params Params, expected []appsv1.St
+}
+
+// hasVolumeClaimsTemplatesChanged if volume claims template change has been detected.
+// We need to this by hand due to some fields being automatically filled by the API server
```
```suggestion
// We need to do this manually due to some fields being automatically filled by the API server
```

jaronoff97 commented on pkg/collector/reconcile/statefulset_test.go:
```diff
@@ -138,7 +139,45 @@ func TestExpectedStatefulsets(t *testing.T) {
+		assert.Equal(t, newSs.Spec.Selector.MatchLabels, actual.Spec.Selector.MatchLabels)
+	})
+
+	t.Run("change Spec.VolumeClaimTemplates should recreate statefulset", func(t *testing.T) {
```
where is the volume claim template changed in this test? It seems we just make a collector once with a volume claim template, but never change it to see the recreation?

matej-g replied:
So we first create STS _with_ volume claim templates and then create a new STS _without_ one, in which case we should also recreate the STS. But it might be confusing for testing purposes, I made it clearer by switching it around (first create STS with no volume claim templates, then add one).

jaronoff97 commented on pkg/collector/volumeclaim.go:
```diff
@@ -19,22 +19,15 @@ import (
-
-	return volumeClaimTemplates
+	// Add all user specified claims.
+	return append([]corev1.PersistentVolumeClaim{}, otelcol.Spec.VolumeClaimTemplates...)
```
Why not just return `otelcol.Spec.VolumeClaimTemplates`?

matej-g replied:
Yes, why not 🤦 , thanks!

lsoica commented:
@matej-g While trying to upgrade the operator to a newer version (that includes this change), I ended up with an infinite reconciliation loop. I have a collector in STS mode, with 2 replicas, without persistent volume(s) and the pods are continuously delete and recreated. The target allocator is enabled on the CR:

```
{"level":"error","ts":"2024-05-07T09:50:40Z","logger":"controllers.OpenTelemetryCollector","msg":"detected immutable field change, trying to delete, new object will be created on next reconcile","opentelemetrycollector":{"name":"metrics","namespace":"updates-ob"},"object_name":"metrics-targetallocator","object_kind":"&TypeMeta{Kind:,APIVersion:,}","existing":"metrics-targetallocator","error":"immutable field change attempted","stacktrace":"github.com/open-telemetry/opentelemetry-operator/controllers.reconcileDesiredObjects\n\t/home/runner/work/opentelemetry-operator/opentelemetry-operator/controllers/common.go:106\ngithub.com/open-telemetry/opentelemetry-operator/controllers.(*OpenTelemetryCollectorReconciler).Reconcile\n\t/home/runner/work/opentelemetry-operator/opentelemetry-operator/controllers/opentelemetrycollector_controller.go:228\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:119\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227"}
```

```
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
spec:
  config: ...
  ingress:
    route: {}
  managementState: managed
  mode: statefulset
  observability:
    metrics: {}
  podDisruptionBudget:
    maxUnavailable: 1
  ports:
  - name: metrics
    port: 4317
    protocol: TCP
    targetPort: 0
  replicas: 2
  resources: {}
  serviceAccount: sa
  targetAllocator:
    enabled: true
    prometheusCR:
      enabled: true
      scrapeInterval: 30s
    replicas: 1
    resources: {}
    serviceAccount: sa
  updateStrategy: {}
  upgradeStrategy: automatic
```

jaronoff97 commented:
@lsoica can you open a new issue for this? Please also upgrade to the latest operator version which may contain fixes.
